test/unit/conditions/condition-expr.test.ts

```diff
@@ -0,0 +1,423 @@
+      );
+      const conditionExprOlderMinorPatchVersion = new ConditionExpression(
+        rpcCondition,
+        '0.9.9'
+      );
+      expect(
+        conditionExprCurrentVersion.equals(conditionExprOlderMinorVersion)
+      ).not.toBeTruthy();
+      expect(
+        conditionExprCurrentVersion.equals(conditionExprOlderPatchVersion)
+      ).not.toBeTruthy();
+      expect(
+        conditionExprCurrentVersion.equals(conditionExprOlderMinorPatchVersion)
+      ).not.toBeTruthy();
+      expect(
+        conditionExprOlderMinorVersion.equals(conditionExprOlderPatchVersion)
+      ).not.toBeTruthy();
+      expect(
+        conditionExprOlderMinorVersion.equals(
+          conditionExprOlderMinorPatchVersion
+        )
+      ).not.toBeTruthy();
+      expect(
+        conditionExprOlderPatchVersion.equals(
+          conditionExprOlderMinorPatchVersion
+        )
+      ).not.toBeTruthy();
+    });
+
+    it.each([
+      erc721BalanceCondition,
+      contractConditionNoAbi,
+      contractConditionWithAbi,
+      timeCondition,
+      compoundCondition,
+    ])('same version but different condition', async (condition) => {
+      const conditionExprSameVersionDifferentCondition =
+        new ConditionExpression(condition);
+      expect(
+        conditionExprCurrentVersion.equals(
+          conditionExprSameVersionDifferentCondition
+        )
+      ).not.toBeTruthy();
+    });
+
+    it('same contract condition although using erc721 helper', async () => {
+      const erc721ConditionExpr = new ConditionExpression(
+        erc721BalanceCondition
+      );
+      const erc721ConditionData = erc721BalanceCondition.toObj();
+      const sameContractCondition = new ContractCondition(erc721ConditionData);
+      const contractConditionExpr = new ConditionExpression(
+        sameContractCondition
+      );
+      expect(erc721ConditionExpr.equals(contractConditionExpr)).toBeTruthy();
+    });
+  });
+
+  describe('serialization / deserialization', () => {
+    it.each([
+      erc721BalanceCondition,
+      contractConditionNoAbi,
+      contractConditionWithAbi,
+      rpcCondition,
+      timeCondition,
+      compoundCondition,
+    ])('serializes to and from json', async (condition) => {
+      const conditionExpr = new ConditionExpression(condition);
+      const conditionExprJson = conditionExpr.toJson();
+      expect(conditionExprJson).toBeDefined();
+      expect(conditionExprJson).toContain('version');
+      expect(conditionExprJson).toContain(ConditionExpression.VERSION);
+      expect(conditionExprJson).toContain('condition');
+      expect(conditionExprJson).toContain(toJSON(condition.toObj()));
+
+      const conditionExprFromJson =
+        ConditionExpression.fromJSON(conditionExprJson);
+      expect(conditionExprFromJson).toBeDefined();
+      expect(conditionExprFromJson.equals(conditionExprFromJson)).toBeTruthy();
+    });
+
+    it('incompatible version', async () => {
+      const invalidVersion = '100.0.0';
```

Comment:
```suggestion
      const invalidVersion = '0.1.0';
```

Admittedly a nitpick, but there is a possibility that, someday, we will reach this version `100.0.0` and then this test will fail.

Using a version lower than the first version (`1.0.0` -> `0.1.0`) ensures that this will never fail.

WDYT?

Reply:
The goal of the test is for the major version to be higher than the current version; hence semver incompatibility and the exception thrown. 

A minor version change is not an incompatibility problem since, based on semver, the change would have been made in a backwards-compatible fashion (https://semver.org/).

I don't really expect us to reach version `100.0.0`, but your underlying point about maintainability is a good one 👍 . I've updated the test to increment the major version based on the current current version.

Reply:
oh, got it. So, yes, that's a good approach.